### ISSUE TITLE
Feature Request: Allow identifying the current async context in some way

### ISSUE THREAD

**Issue by Blackclaws** [Cysharp/UniTask#455]
So I understand why ExecutionContext is not actually taken into consideration when using UniTask, and while it would be nice to have it at least as an opt-in to help with debugging (or for specific calls, similar to ConfigureAwait()) that is probably not going to happen.

What would be nice to have is some sort of property that allows you to identify the current async context in some way.

To give an example of when this would be useful:

We have the very real situation of a reentrant async method that can be called from multiple places but isn't Thread safe. Therefore on entry to this method we block on a SemaphoreSlim to make sure that only one call at a time takes place. However there is the possibility of reentrancy within this method, i.e. it can be indirectly recursive. In this case it would be nice if an async child could enter that method while it is blocked on an await call. To make this possible I'd have to identify whether the entering call is part of the same async chain as the one currently holding the lock. 

Right now I have no way to identify the async stack. Would this be something that you would consider adding or is that too heavyweight?

**Comment by timcassell** [comment]
Unity still uses the old .Net Framework/Mono implementation where `ExecutionContext` is quite heavy. It allocates when it is captured, even if it is not written to. Using that would make UniTask methods a lot more expensive. .Net Core/.Net 5+ optimized it so it no longer allocates when it is captured and is much faster, it only allocates when you write to it, but it still has some overhead. It might be worth it when Unity finally upgrades the runtime.

If you want to try it out, I wrote a similar library that does support `ExecutionContext`, but you have to enable it explicitly (because of said overhead). https://github.com/timcassell/ProtoPromise#configuration

**Comment by timcassell** [comment]
> Interesting, it seems to be quite different from standard Task semantics though and we're already heavily invested in UniTasks across the whole application so not likely we're going to use it at least for this project.

The semantics are the same as UniTask, it's just different naming. But I understand not wanting to switch.

> Yeah I understand why you would not want to capture the whole ExecutionContext, though there should be a possibility to trace async operations. I mean UniTask already supports Stack Traces in the TaskTracker, whihc of course has its own performance implications, but in principle its possible, the overhead would have to be evaluated.

You would have to add `ExecutionContext` support to the library. I don't think there is any other way. Since you have access to the source code, you could look at how I implemented it in mine and do the same in UniTask (look at the async method builders).

**Comment by hadashiA** [comment]
we've considered it, but as timecassell says, don't see any other way than to pass on something like ExecutionContext.
But we are not going to do that, so I don't think it will be supported.
